packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts

```diff
@@ -0,0 +1,2266 @@
+        viewport.renderFrame();
+        await assertCategoryVisibility({ categoryIdFilter: categoryId, perModelVisibilityOverride: true, handlerVisibility: "partial" });
+      });
+    });
+
+    describe("element", () => {
+      let modelId: string;
+      let categoryId: string;
+      let elementId: string;
+      let childElementId: string;
+
+      before(() => {
+        modelId = firstModelId;
+        categoryId = getFirstValue(models.get(modelId)!.values());
+        elementId = getFirstValue(categoryParentElements.get(categoryId)!.values());
+        childElementId = getFirstValue(elementHierarchy.get(elementId)!.values());
+      });
+
+      it("if model is hidden, showing element makes model and category partially visible", async () => {
+        const elementNode = createElementHierarchyNode({ modelId, categoryId, hasChildren: true, elementId });
+        await handler.changeVisibility(elementNode, true);
+        expect(viewport.alwaysDrawn).to.deep.eq(new Set([elementId]));
+        viewport.renderFrame();
+        await Promise.all([
+          assertModelVisibility({
+            modelId,
+            viewportVisibility: true,
+            handlerVisibility: "partial",
+          }),
+          assertCategoryVisibility({
+            categoryIdFilter: categoryId,
+            handlerVisibility: "partial",
+          }),
+          assertCategoryVisibility({
+            modelIdFilter: modelId,
+            categoryIdFilter: (id) => id !== categoryId,
+            handlerVisibility: "hidden",
+          }),
+          assertElementsVisibility({
+            elementIdFilter: elementId,
+            visibility: "visible",
+          }),
+        ]);
+      });
+
+      it("if model is hidden, showing element removes all model elements from the always drawn list", async () => {
+        const elementIds = categoryParentElements.get(categoryId);
+        viewport.setAlwaysDrawn(new Set(elementIds));
+
+        const elementNode = createElementHierarchyNode({ modelId, categoryId, hasChildren: true, elementId });
+        await handler.changeVisibility(elementNode, true);
+
+        expect(viewport.alwaysDrawn).to.deep.eq(new Set([elementId]));
+        viewport.renderFrame();
+        await Promise.all([
+          assertModelVisibility({
+            modelId,
+            viewportVisibility: true,
+            handlerVisibility: "partial",
+          }),
+          assertCategoryVisibility({
+            categoryIdFilter: categoryId,
+            handlerVisibility: "partial",
+          }),
+          assertCategoryVisibility({
+            modelIdFilter: modelId,
+            categoryIdFilter: (id) => id !== categoryId,
+            handlerVisibility: "hidden",
+          }),
+          assertElementsVisibility({
+            elementIdFilter: elementId,
+            visibility: "visible",
+          }),
+        ]);
+      });
+
+      it("hiding parent element makes all children hidden", async () => {
```

Comment:
This should not be happening, but the test, I assume, passes. I guess the elements are hidden because the category is hidden? Would the test fail if you didn't set the parent element's visibility to "hide"?

Reply:
This was a bug in `assertElementVisibility`, which made it always pass. Fixed the issue and adjusted the element tests to the new requirements